# Sistemas Inteligentes 2021/2022

## Mini-projeto 2: Quadrados Latinos


## Grupo: 42

### Elementos do Grupo

Número:  53687       Nome:   Ariana Dias  
Número:  53746       Nome:   Andrei Tataru  
Número:  51127       Nome:   Luís Ferreirinha  

(Nota: Neste relatório pode adicionar as células de texto e código que achar necessárias.)

## Representação de variáveis, domínios, vizinhos e restrições

O problema dos quadrados latinos pode ser representado como uma matriz $n \times n$ em que cada linha e coluna contêm exatmente um número de $1$ a $n$.  

Então para caracterizar este CSP em python obtamos pelas seguintes escolhas:  

- Variáveis:  
  
    Cada variável no nosso problema vai representar o valor atribuiado a um elemento da matriz.  
    Então em python representamos o conjunto das variáveis como uma lista das coordenadas de cada
    elemento na matriz:    
    -  var $=[(i_1,j_1),...,(i_n,j_n)]$  
<br>
- Domínios:  
  
    Neste problema cada variável pode tomar um valor de $1,...,n$, exceto de houver condições iniciais.
    Portanto em python decidimos utilizar um dicionário, onde as chaves são as variáveis e estas correspondem a uma lista de inteiros contento os valores possíveis para aquela variável:
    
    - dominio $=\{(i_1,j_1) = [1,..,n], ...\}$  
<br> 
- Vizinhos:  

    Uma variável vai ser vizinha de outra de estiver na mesma linha ou coluna que essa variável.
    Em python definimos os vizinhos como um dicionário onde cada variável é uma chave e corresponde a uma lista de variáveis que são suas vizinhas:  
    
    - vizinhos $=\{(i_k,i_m) : [(i_1,j_m),...,(i_n,j_m),(i_k,j_1),...,(i_k,j_n)], \, ...  \, \}$  
<br>
- Restrições:  

    A restrição principal deste problema é a impossibilidade de haver repetições de um número
    ao longo de uma coluna ou linha, ou seja tem de ser diferente de todos os seus vizinhos.
    Em python esta restrição é representada pela comparação dos valores de duas variáveis vizinhas:  

    -  ```python
       valor_variavel_1 != valor_variavel_2
       ```
    <br>
    Como também temos a variação do Futoshiki, temos de definir outras restrições para esse problema.  

    A ideia principal é que duas variáveis adjacentes podem agora ter a restrição adicional de o valor de uma ser maior ou menor que o valor da outra. Então para representar o mapeamento destas novas restrições usamos um dicionário onde as chaves são um tuplos contendos as duas variáveis em questão e vão corresponde a uma função que verifica se os valores destas são menores ou maiores conforme a restrição.

    - desigualdades $=\{((i_k,j_m),(i_k,j_{m+1})) : maior\, \, \text{ou} \, \, menor \}$  

    Cada uma destas funções vair comparar os valores das variavéis conforme o seu nome.  

    - ```python 
      def maior(a,b): return a > b
      ```

    - ```python 
      def menor(a,b): return a < b
      ```


 

## Formulação do problema

In [1]:
from csp import *

def quadrado_latino(n=3, initial={}) -> CSP:
    """Gera um CSP para o problema dos quadrados latinos

    Args:
        n (int, optional): Dimensao da matriz do problema. Defaults to 3.
        initial (dict, optional): Valores iniciais para cada um dos elementos da matriz. Defaults to {}.

    Returns:
        CSP: Objeto CSP para o problema dos quadrados latinos
    """    
    
    # definicao das variaveis, as coordenadas x,y vão de 0 a n-1
    variaveis = [(i,j) for i in range(0,n) for j in range(0,n)]

    # Valores possiveis para as variaveis dada a dimensao n do problema
    conjunto = [i for i in range(1,n+1)]
    
    # Iniciamos os dominios com os valores iniciais dados para as variaveis
    dominios = initial

    # Diferenca dos dois conjuntos vai resultar nas variaveis nao preenchidas
    for item in set(variaveis - initial.keys()):
        dominios[item] = conjunto
    
    # Condicao que determina de duas variaveis sao vizinhas, verifica se tem alguma das coordendas x ou y iguais
    condicao_vizinhanca = lambda x,y : (x[0] == y[0] and x[1] != y[1]) or (x[0] != y[0] and x[1] == y[1])

    # Para cada variavel vamos atribuir uma lista resultante de filtrar a lista de todas as variaveis
    # com a condicao de que as variaveis sejam vizinhas da variavel chave
    vizinhos = {var : list(filter(lambda x: condicao_vizinhanca(x,var), variaveis)) for var in variaveis}

    def restricoes(X, a, Y, b):
        # Verifica-se se duas variaveis sao vizinhas antes de verficar que tem valores diferentes entre si
        if Y in vizinhos[X]:
            return a != b
        return True
        
    return CSP(variaveis, dominios, vizinhos, restricoes)

## Criação do problema do quadrado latino simples

Mostrem que o código está a funcionar, construindo um problema de quadrado latino *4x4*, imprimindo as variáveis, domínios iniciais, e vizinhos. Adicione os comentários necessários. Mostre como podemos criar um puzzle com quadrados já preenchidos, e qual o impacto que isso tem nas variáveis, domínios iniciais, e vizinhos.

Para criar um problema quadrado latino 4x4 sem valores iniciais para as variáveis basta executar

In [2]:
q4 = quadrado_latino(4)

Tendo o problema definido, podemos agora ver as variáveis, domínios e os vizinhos deste problema

In [3]:
print(f"Variáveis = {q4.variables}\n")
print(f"Domínios = {q4.domains}\n")
print(f"Vizinhos = {q4.neighbors}\n")

Variáveis = [(0, 0), (0, 1), (0, 2), (0, 3), (1, 0), (1, 1), (1, 2), (1, 3), (2, 0), (2, 1), (2, 2), (2, 3), (3, 0), (3, 1), (3, 2), (3, 3)]

Domínios = {(0, 1): [1, 2, 3, 4], (1, 2): [1, 2, 3, 4], (2, 1): [1, 2, 3, 4], (3, 1): [1, 2, 3, 4], (0, 2): [1, 2, 3, 4], (2, 2): [1, 2, 3, 4], (1, 0): [1, 2, 3, 4], (3, 2): [1, 2, 3, 4], (1, 3): [1, 2, 3, 4], (0, 0): [1, 2, 3, 4], (1, 1): [1, 2, 3, 4], (0, 3): [1, 2, 3, 4], (2, 0): [1, 2, 3, 4], (3, 0): [1, 2, 3, 4], (2, 3): [1, 2, 3, 4], (3, 3): [1, 2, 3, 4]}

Vizinhos = {(0, 0): [(0, 1), (0, 2), (0, 3), (1, 0), (2, 0), (3, 0)], (0, 1): [(0, 0), (0, 2), (0, 3), (1, 1), (2, 1), (3, 1)], (0, 2): [(0, 0), (0, 1), (0, 3), (1, 2), (2, 2), (3, 2)], (0, 3): [(0, 0), (0, 1), (0, 2), (1, 3), (2, 3), (3, 3)], (1, 0): [(0, 0), (1, 1), (1, 2), (1, 3), (2, 0), (3, 0)], (1, 1): [(0, 1), (1, 0), (1, 2), (1, 3), (2, 1), (3, 1)], (1, 2): [(0, 2), (1, 0), (1, 1), (1, 3), (2, 2), (3, 2)], (1, 3): [(0, 3), (1, 0), (1, 1), (1, 2), (2, 3), (3, 3)], (2, 0): [(0, 0), 

Para criar-mos um problema com um quadrado semi-preenchido temos primeiro de definir quais os elementos da matriz que vão estar preenchidos e qual o seu valor. Para fazer-mos isso podemos definir um dicionário com essa informação da seguinte forma:

In [4]:
valores_preenchidos = {(0,0) : [1], (1,1) : [3], (2,2) : [2]}

Depois basta passar esse dicionário como segundo argumento à função quadrado_latino

In [5]:
q4_semi = quadrado_latino(n=4, initial=valores_preenchidos)

O facto de o quadrado ser semi preenchido apenas vai afetar os domínios iniciais, fazendo com que o dominio inicial das variáveis já com valores atribuidos seja reduzido esses valores.

In [6]:
print(f"Variáveis = {q4_semi.variables}\n")
print(f"Domínios = {q4_semi.domains}\n")
print(f"Vizinhos = {q4_semi.neighbors}\n")

Variáveis = [(0, 0), (0, 1), (0, 2), (0, 3), (1, 0), (1, 1), (1, 2), (1, 3), (2, 0), (2, 1), (2, 2), (2, 3), (3, 0), (3, 1), (3, 2), (3, 3)]

Domínios = {(0, 0): [1], (1, 1): [3], (2, 2): [2], (0, 1): [1, 2, 3, 4], (1, 2): [1, 2, 3, 4], (2, 1): [1, 2, 3, 4], (3, 1): [1, 2, 3, 4], (0, 3): [1, 2, 3, 4], (2, 0): [1, 2, 3, 4], (3, 0): [1, 2, 3, 4], (2, 3): [1, 2, 3, 4], (0, 2): [1, 2, 3, 4], (3, 3): [1, 2, 3, 4], (1, 0): [1, 2, 3, 4], (3, 2): [1, 2, 3, 4], (1, 3): [1, 2, 3, 4]}

Vizinhos = {(0, 0): [(0, 1), (0, 2), (0, 3), (1, 0), (2, 0), (3, 0)], (0, 1): [(0, 0), (0, 2), (0, 3), (1, 1), (2, 1), (3, 1)], (0, 2): [(0, 0), (0, 1), (0, 3), (1, 2), (2, 2), (3, 2)], (0, 3): [(0, 0), (0, 1), (0, 2), (1, 3), (2, 3), (3, 3)], (1, 0): [(0, 0), (1, 1), (1, 2), (1, 3), (2, 0), (3, 0)], (1, 1): [(0, 1), (1, 0), (1, 2), (1, 3), (2, 1), (3, 1)], (1, 2): [(0, 2), (1, 0), (1, 1), (1, 3), (2, 2), (3, 2)], (1, 3): [(0, 3), (1, 0), (1, 1), (1, 2), (2, 3), (3, 3)], (2, 0): [(0, 0), (1, 0), (2, 1), (2, 2), (2,

Resolva o problema com o backtracking sem inferencia, com inferencia, e com uma heurística.

Vamos definir uma função que executa uma pesquisa $n$ vezes (neste caso 50) e calcula a média dos tempos de execução da pesquisa. Desta forma podemos obter tempos de execução mais consistentes que não vão depender tanto de variações na velocidade do processador durante a execução das diferentes procuras.

In [7]:
import timeit

def avg_run_time(func, *args, **kwargs):
    """Corre uma função n vezes medindo o tempo de execução de cada vez
    e imprime a média dos tempos

    Args:
        func: Função a correr
    """
    n = 20
    run_times = []
    for i in range(n):
        start = timeit.default_timer()
        func(*args, **kwargs)
        stop = timeit.default_timer()
        run_times.append(stop-start)
    return sum(run_times) / n

# Guardar os dados e solucoes para analisar mais tarde
tempos = {}
solucoes = {}

- Dimensão do problema:

In [8]:
n = 4

- Backtracking sem inferência

In [9]:
# Definir problema
QuadradoLatinoSemInferência = quadrado_latino(n)
# Guardar solucao
sol_1 = backtracking_search(QuadradoLatinoSemInferência)
# Calcular tempo medio executando 20 vezes
tempo_medio1 = avg_run_time(backtracking_search, QuadradoLatinoSemInferência)
# Guardar resultados
tempos['QuadradoLatinoSemInferência'] = tempo_medio1
solucoes['QuadradoLatinoSemInferência'] = sol_1
# Imprimir resultados
print(f"Tempo médio de execução: {tempo_medio1}")
print('Assignment (QuadradoLatinoSemInferência) = ',sol_1)

Tempo médio de execução: 0.00033662999339867383
Assignment (QuadradoLatinoSemInferência) =  {(0, 0): 1, (0, 1): 2, (0, 2): 3, (0, 3): 4, (1, 0): 2, (1, 1): 1, (1, 2): 4, (1, 3): 3, (2, 0): 3, (2, 1): 4, (2, 2): 1, (2, 3): 2, (3, 0): 4, (3, 1): 3, (3, 2): 2, (3, 3): 1}


- Backtracking com inferência Foward-Checking

In [10]:
# Definir problema
QuadradoLatinoComInferência_FRWDchecking = quadrado_latino(n)
# Resolver e guardar solucao
sol_2 = backtracking_search(QuadradoLatinoComInferência_FRWDchecking, inference=forward_checking)
# Calcular o tempo medio de execucao
tempo_medio2 = avg_run_time(backtracking_search, QuadradoLatinoComInferência_FRWDchecking, inference=forward_checking)
# Guardar resultados
tempos['QuadradoLatinoComInferência_FRWDchecking'] = tempo_medio2
solucoes['QuadradoLatinoComInferência_FRWDchecking'] = sol_2
# Imprimir resultados
print(f"Tempo médio de execução: {tempo_medio2}")
print('Assignment (QuadradoLatinoComInferência_FRWDchecking) = ',sol_2)

Tempo médio de execução: 0.00021894499950576575
Assignment (QuadradoLatinoComInferência_FRWDchecking) =  {(0, 0): 1, (0, 1): 2, (0, 2): 3, (0, 3): 4, (1, 0): 2, (1, 1): 1, (1, 2): 4, (1, 3): 3, (2, 0): 3, (2, 1): 4, (2, 2): 1, (2, 3): 2, (3, 0): 4, (3, 1): 3, (3, 2): 2, (3, 3): 1}


- Backtracking com inferência MAC

In [11]:
# Definir problema
QuadradoLatinoComInferência_MAC = quadrado_latino(n)
# Resolver e guardar solucao
sol_4 = backtracking_search(QuadradoLatinoComInferência_MAC, inference=mac)
# Calcular tempo medio de execucao
tempo_medio4 = avg_run_time(backtracking_search, QuadradoLatinoComInferência_MAC, inference=mac)
# Guarda resultados
tempos['QuadradoLatinoComInferência_MAC'] = tempo_medio4
solucoes['QuadradoLatinoComInferência_MAC'] = sol_4
# Imprimir resultados
print(f"Tempo médio de execução: {tempo_medio4}")
print('Assignment (QuadradoLatinoComInferência_MAC) = ', sol_4)

Tempo médio de execução: 0.0006403200008207932
Assignment (QuadradoLatinoComInferência_MAC) =  {(0, 0): 1, (0, 1): 2, (0, 2): 3, (0, 3): 4, (1, 0): 2, (1, 1): 1, (1, 2): 4, (1, 3): 3, (2, 0): 3, (2, 1): 4, (2, 2): 1, (2, 3): 2, (3, 0): 4, (3, 1): 3, (3, 2): 2, (3, 3): 1}


- Backtracking com pré-processamento AC3

In [12]:
# Definir problema
QuadradoLatinoComAC3 = quadrado_latino(n)
# Aplicar pre-processamento
p_AC3 = AC3(QuadradoLatinoComAC3)
# Resolver e guardar solucao
sol_3 = backtracking_search(p_AC3)
# Calcular o tempo medio de execucao
tempo_medio3 = avg_run_time(backtracking_search, p_AC3)
# Guarda resultados
tempos['QuadradoLatinoComAC3'] = tempo_medio3
solucoes['QuadradoLatinoComAC3'] = sol_3
# Imprimir resultados
print(f"Tempo médio de execução: {tempo_medio3}")
print('Assignment (QuadradoLatinoComAC3) = ', sol_3)

Tempo médio de execução: 0.0003123099886579439
Assignment (QuadradoLatinoComAC3) =  {(0, 0): 1, (0, 1): 2, (0, 2): 3, (0, 3): 4, (1, 0): 2, (1, 1): 1, (1, 2): 4, (1, 3): 3, (2, 0): 3, (2, 1): 4, (2, 2): 1, (2, 3): 2, (3, 0): 4, (3, 1): 3, (3, 2): 2, (3, 3): 1}


- Backtracking com Heurística MRV

In [13]:
# Definir problema
QuadradoLatinoComHeuristica = quadrado_latino(n)
# Resolver e guardar
sol_6 = backtracking_search(QuadradoLatinoComHeuristica, select_unassigned_variable = mrv)
# Tempo medio
tempo_medio6 = avg_run_time(backtracking_search, QuadradoLatinoComHeuristica, select_unassigned_variable = mrv)
# Guarda resultados
tempos['QuadradoLatinoComHeuristica'] = tempo_medio6
solucoes['QuadradoLatinoComHeuristica'] = sol_6
# Imprimir resultados
print(f"Tempo médio de execução: {tempo_medio6}")
print('Assignment (QuadradoLatinoComHeuristica) = ', sol_6)

Tempo médio de execução: 0.0005113949999213218
Assignment (QuadradoLatinoComHeuristica) =  {(1, 0): 1, (2, 0): 2, (3, 3): 1, (3, 2): 2, (1, 2): 3, (2, 2): 1, (0, 2): 4, (0, 0): 3, (2, 1): 4, (3, 0): 4, (1, 1): 2, (0, 3): 2, (3, 1): 3, (2, 3): 3, (1, 3): 4, (0, 1): 1}


- Backtracking com Heurística MRV, Inferência MAC e pré-processamento AC3

In [14]:
# Definir problema
QuadradoLatinoHeuristicaInferenciaAC3 = quadrado_latino(n)
# Pre processamento
p_AC3 = AC3(QuadradoLatinoHeuristicaInferenciaAC3)
# Resolver e guardar
sol_7 = backtracking_search(p_AC3, select_unassigned_variable=mrv, inference=mac)
# Tempo medio
tempo_medio7 = avg_run_time(backtracking_search, p_AC3, select_unassigned_variable=mrv, inference=mac)
# Guarda resultados
tempos['QuadradoLatinoHeuristicaInferenciaAC3'] = tempo_medio7
solucoes['QuadradoLatinoHeuristicaInferenciaAC3'] = sol_7
# Imprimir resultados
print(f"Tempo médio de execução: {tempo_medio7}")
print('Assignment (QuadradoLatinoHeuristicaInferenciaAC3) = ', sol_7)

Tempo médio de execução: 0.0006232399988221005
Assignment (QuadradoLatinoHeuristicaInferenciaAC3) =  {(2, 2): 1, (3, 2): 2, (1, 2): 3, (0, 2): 4, (3, 3): 1, (3, 1): 3, (3, 0): 4, (1, 3): 2, (0, 3): 3, (1, 0): 1, (2, 3): 4, (2, 0): 3, (2, 1): 2, (1, 1): 4, (0, 1): 1, (0, 0): 2}


In [15]:
import pandas as pd
temposSorted = dict(sorted(tempos.items(), key=lambda x:x[1]))
pd.DataFrame(temposSorted.values(),index = temposSorted.keys(), columns= [n])


4
QuadradoLatinoComInferência_FRWDchecking  0.000219
QuadradoLatinoComAC3                      0.000312
QuadradoLatinoSemInferência               0.000337
QuadradoLatinoComHeuristica               0.000511
QuadradoLatinoHeuristicaInferenciaAC3     0.000623
QuadradoLatinoComInferência_MAC           0.000640

## Conclusões sobre os resultados temporais
_(todos os valores são referentes a médias)_
Com a tabela acima é possível concluir que:
- o algoritmo que demora menos tempo a encontrar uma solução é a **procura com Retrocesso usando a inferência Forward Checking**, demora apenas 0.000219 segundos;
- usando procura com retrocesso com inferência, verifica-se que **Forward Checking** é mais eficiente que **MAC**;
- a solução do problema utilizando a procura com retrocesso sem inferências nem heurísticas demora menos tempo (0.000337 segundos) que a procura com retrocesso com inferência e heurisrica após pré-processamento com recurso ao algoritmo **AC3** (0.000623 segundos), quase o dobro do tempo;


Acho que como a dimensao deste problema é muito pequena, a procura com retrocesso normal acaba por ser a mais rapida porque vai realizar menos verificações ao longo da execução, e o custo de fazer essas verificações vai ser maior que o custo de voltar a atrás mais vezes na procura porque este problema é muito pequeno, se fosse maior iamos ver mais diferenças.

In [17]:
sols = pd.DataFrame(solucoes).T
sols

0           1           2           \
                                          0  1  2  3  0  1  2  3  0  1  2  3   
QuadradoLatinoSemInferência               1  2  3  4  2  1  4  3  3  4  1  2   
QuadradoLatinoComInferência_FRWDchecking  1  2  3  4  2  1  4  3  3  4  1  2   
QuadradoLatinoComInferência_MAC           1  2  3  4  2  1  4  3  3  4  1  2   
QuadradoLatinoComAC3                      1  2  3  4  2  1  4  3  3  4  1  2   
QuadradoLatinoComHeuristica               3  1  4  2  1  2  3  4  2  4  1  3   
QuadradoLatinoHeuristicaInferenciaAC3     2  1  4  3  1  4  3  2  3  2  1  4   

                                          3           
                                          0  1  2  3  
QuadradoLatinoSemInferência               4  3  2  1  
QuadradoLatinoComInferência_FRWDchecking  4  3  2  1  
QuadradoLatinoComInferência_MAC           4  3  2  1  
QuadradoLatinoComAC3                      4  3  2  1  
QuadradoLatinoComHeuristica               4  3  2  1  
QuadradoLatinoHeuristicaInferenciaAC3     4  3  2  1

In [21]:
sols.duplicated()

QuadradoLatinoSemInferência                 False
QuadradoLatinoComInferência_FRWDchecking     True
QuadradoLatinoComInferência_MAC              True
QuadradoLatinoComAC3                         True
QuadradoLatinoComHeuristica                 False
QuadradoLatinoHeuristicaInferenciaAC3       False
dtype: bool

In [22]:
sols.drop_duplicates()

0           1           2           3  \
                                       0  1  2  3  0  1  2  3  0  1  2  3  0   
QuadradoLatinoSemInferência            1  2  3  4  2  1  4  3  3  4  1  2  4   
QuadradoLatinoComHeuristica            3  1  4  2  1  2  3  4  2  4  1  3  4   
QuadradoLatinoHeuristicaInferenciaAC3  2  1  4  3  1  4  3  2  3  2  1  4  4   

                                                
                                       1  2  3  
QuadradoLatinoSemInferência            3  2  1  
QuadradoLatinoComHeuristica            3  2  1  
QuadradoLatinoHeuristicaInferenciaAC3  3  2  1

## Conclusões sobre os resultados de soluções

Podemos concluir que usando a heurística MRV a solução do problema é diferente quando não a usamos. Mais ainda, verifica-se que mesmo usando a heurística MRV, a solução do problema é diferente quando se fez um pré-processamento com recurso ao algoritmo AC3.



## Visualização do problema

In [23]:
def visualizacao_quadrado_latino(n, problema_csp, solucao):
    """Gera uma visualização do estado inicial e da solução de um problema de quadrados
    latinos de dimensão n
    
    Args:
        n (int): Dimensão do problema
        problema_csp (CSP): Object CSP do problema em questão
        solucao (dict): Solução do problema
    """    
    
    estado_inicial = [[0 for y in range(n)] for x in range(n)]
    for var, dom in problema_csp.domains.items():
        if len(dom) == 1:
            estado_inicial[var[1]][var[0]] = dom[0]
        
    estado_final = [[0 for y in range(n)] for x in range(n)]
    for var, valor in solucao.items():
        estado_final[var[1]][var[0]] = valor
    
    board_inicial = ""
    board_final = ""
    for y in range(n):
        for x in range(n):
            board_inicial += str(estado_inicial[y][x]) + " "
            board_final += str(estado_final[y][x]) + " "
        board_inicial += "\n"
        board_final += "\n"
    
    print("Estado Inicial:")
    print(board_inicial)
    print("Solução:")
    print(board_final)

In [24]:
visualizacao_quadrado_latino(4, quadrado_latino(4), backtracking_search(quadrado_latino(4)))

Estado Inicial:
0 0 0 0 
0 0 0 0 
0 0 0 0 
0 0 0 0 

Solução:
1 2 3 4 
2 1 4 3 
3 4 1 2 
4 3 2 1 



## Criação do problema Futoshiki *5x5*

Mostrem que o código está a funcionar, construindo um problema de Futoshiki *5x5*, imprimindo as variáveis, domínios iniciais, e vizinhos. Adicione os comentários necessários. Utilize o [link](https://www.futoshiki.org/) para gerar puzzles e validar a implementação.

In [25]:
def maior(x,y):
    return x > y

def menor(x,y):
    return x < y

def futoshiki(n=3, initial={}, desigualdades={}):
    """
    Pode receber parametros ou não.
    Deve devolver um CSP, à semelhança dos guiões das aulas PL.
    Comente o código.
    """
    variaveis = [(i,j) for j in range(0,n) for i in range(0,n)]

    conjunto = [i for i in range(1,n+1)]
    
    dominios = initial

    # Diferenca dos dois conjuntos vai resultar nas variaveis nao preenchidas
    for item in set(variaveis - initial.keys()):
        dominios[item] = conjunto
    
    condicao_vizinhanca = lambda x,y : ((x[0] == y[0] and x[1] != y[1]) or (x[0] != y[0] and x[1] == y[1]))
    
    vizinhos = {var : list(filter(lambda x: condicao_vizinhanca(x,var), variaveis)) for var in variaveis}

    def restricoes(X, a, Y, b):
        if (X,Y) in desigualdades.keys():
            if not desigualdades[(X,Y)](a,b):
                return False
        if (Y,X) in desigualdades.keys():
            if not desigualdades[(Y,X)](b,a):
                return False
        if Y in vizinhos[X]:
            return  a != b
        return True

    return CSP(variaveis, dominios, vizinhos, restricoes)

In [26]:
preenchidos = {(1,3) : [1]}
desigualdades = {((1,0),(2,0)) : maior, ((2,0),(3,0)) : maior, ((1,1),(1,0)): maior, ((3,3),(3,2)) : maior}

In [27]:
p = futoshiki(4, preenchidos, desigualdades)
print("Variáveis = ", p.variables)
print("Domínios = ", p.domains)
print("Vizinhos = ", p.neighbors)
r = backtracking_search(p)

Variáveis =  [(0, 0), (1, 0), (2, 0), (3, 0), (0, 1), (1, 1), (2, 1), (3, 1), (0, 2), (1, 2), (2, 2), (3, 2), (0, 3), (1, 3), (2, 3), (3, 3)]
Domínios =  {(1, 3): [1], (0, 1): [1, 2, 3, 4], (1, 2): [1, 2, 3, 4], (2, 1): [1, 2, 3, 4], (0, 0): [1, 2, 3, 4], (3, 1): [1, 2, 3, 4], (1, 1): [1, 2, 3, 4], (0, 3): [1, 2, 3, 4], (2, 0): [1, 2, 3, 4], (3, 0): [1, 2, 3, 4], (2, 3): [1, 2, 3, 4], (0, 2): [1, 2, 3, 4], (3, 3): [1, 2, 3, 4], (2, 2): [1, 2, 3, 4], (1, 0): [1, 2, 3, 4], (3, 2): [1, 2, 3, 4]}
Vizinhos =  {(0, 0): [(1, 0), (2, 0), (3, 0), (0, 1), (0, 2), (0, 3)], (1, 0): [(0, 0), (2, 0), (3, 0), (1, 1), (1, 2), (1, 3)], (2, 0): [(0, 0), (1, 0), (3, 0), (2, 1), (2, 2), (2, 3)], (3, 0): [(0, 0), (1, 0), (2, 0), (3, 1), (3, 2), (3, 3)], (0, 1): [(0, 0), (1, 1), (2, 1), (3, 1), (0, 2), (0, 3)], (1, 1): [(1, 0), (0, 1), (2, 1), (3, 1), (1, 2), (1, 3)], (2, 1): [(2, 0), (0, 1), (1, 1), (3, 1), (2, 2), (2, 3)], (3, 1): [(3, 0), (0, 1), (1, 1), (2, 1), (3, 2), (3, 3)], (0, 2): [(0, 0), (0, 1), 

In [28]:
print(r)

{(0, 0): 4, (1, 0): 3, (2, 0): 2, (3, 0): 1, (0, 1): 3, (1, 1): 4, (2, 1): 1, (3, 1): 2, (0, 2): 1, (1, 2): 2, (2, 2): 4, (3, 2): 3, (0, 3): 2, (1, 3): 1, (2, 3): 3, (3, 3): 4}


Resolva o problema com o backtracking sem inferencia, com inferencia, e com uma heurística. Até que dimensão consegue resolver o problema em menos de 1 minuto?

#### Problema Futoshiki *5x5*

In [122]:
preenchidos5={(0,0):[3], (1,4):[2], (4,4):[1]}

desigualdades5 = { ((0,0),(1,0)) : maior, ((2,0),(3,0)) : menor,((1,2),(2,2)) : menor , ((2,2),(3,2)) : menor, ((0,4),(0,3)) : maior}
futoshiki5 = futoshiki(5, preenchidos5, desigualdades5)

#### Problema Futoshiki *6x6*

In [123]:
preenchidos6 = {(4,3): [1], (5,4): [2]}
desigualdades6 = {((0,0),(1,0)):menor, ((1,0),(1,1)):menor, ((0,2),(1,2)):maior, ((3,0),(3,1)):maior, ((3,2),(3,3)):menor, ((2,4),(3,4)):maior, ((3,3),(3,4)):menor, ((3,0),(4,0)):menor, ((4,0),(4,1)):menor, ((4,1),(4,2)):menor, ((5,0),(5,1)):maior, ((5,1),(5,2)):maior}
futoshiki6 = futoshiki(6, preenchidos6, desigualdades6)

#### Problema Futoshiki *7x7*

In [124]:
preenchidos7 = {(1,3): [5], (5,2): [5]}
desigualdades7 = { ((0,0),(1,0)) : menor, ((0,1),(0,0)) : menor, ((4,1),(4,0)) : menor,((4,1),(5,1)) : maior, ((6,1),(6,0)) : menor, ((0,2),(1,2)) : maior, ((2,2),(3,2)) : menor, ((5,2),(6,2)) : maior, ((3,3),(3,2)) : maior, ((6,3),(6,2)) : maior, ((1,3),(2,3)) : maior, ((0,4),(0,3)) : maior, ((4,4),(4,3)) : menor, ((1,4),(2,4)) : maior, ((1,5),(1,4)) : maior, ((4,5),(4,4)) : menor, ((4,6),(4,5)) : maior, ((6,5),(6,4)) : menor, ((0,6),(1,6)) : maior, ((2,6),(3,6)) : maior}
futoshiki7 = futoshiki(7, preenchidos7, desigualdades7)

##### extreme

In [125]:
preenchidos7ext = {(3,2) : [2], (4,0) : [4]}
desigualdades7ext = {((1,0),(2,0)):menor, ((2,0),(3,0)):menor, ((5,0),(6,0)): maior, ((2,1),(2,0)):menor, ((5,1),(5,0)):menor, ((0,1),(1,1)):maior, ((1,1),(2,1)):maior, ((2,1),(3,1)): menor, ((4,1),(5,1)):menor, ((2,2),(2,1)): menor, ((5,2),(5,1)): maior, ((1,3),(1,2)): menor, ((2,3),(2,2)): maior, ((0,3),(1,3)):maior, ((2,3),(3,3)):menor, ((4,3),(5,3)): menor, ((0,4),(0,3)):maior,((1,4),(1,3)):menor,((6,4),(6,3)):menor, ((3,4),(4,4)):menor,((4,4),(5,4)):menor,((3,6),(3,5)):maior, ((3,5),(3,4)):menor, ((0,6),(0,5)): maior, ((0,2),(1,2)): maior}

futoshiki7Extreme = futoshiki(7, preenchidos7ext, desigualdades7ext)

### Aplicação de Algoritmos

In [127]:
solucoesFutoshiki = {}
temposFutoshiki = {}

- Backtracking sem inferência

In [135]:
# Definir problema
Futoshiki5SemInferência = futoshiki(5, preenchidos5, desigualdades5)
# Guardar solucao
sol_15 = backtracking_search(Futoshiki5SemInferência)
# Calcular tempo medio executando 20 vezes
tempo_medio15 = avg_run_time(backtracking_search,Futoshiki5SemInferência)
# Guardar resultados
temposFutoshiki['Futoshiki5SemInferência'] = tempo_medio15
solucoesFutoshiki['Futoshiki5SemInferência'] = sol_15
# Imprimir resultados
print(f"Tempo médio de execução: {tempo_medio15}")
print('Assignment (Futoshiki5SemInferência) = ',sol_15)

# Definir problema
Futoshiki6SemInferência = futoshiki(6, preenchidos6, desigualdades6)
# Guardar solucao
sol_16 = backtracking_search(Futoshiki6SemInferência)
# Calcular tempo medio executando 20 vezes
tempo_medio16 = avg_run_time(backtracking_search,Futoshiki6SemInferência)
# Guardar resultados
temposFutoshiki['Futoshiki6SemInferência'] = tempo_medio16
solucoesFutoshiki['Futoshiki6SemInferência'] = sol_16
# Imprimir resultados
print(f"Tempo médio de execução: {tempo_medio16}")
print('Assignment (Futoshiki6SemInferência) = ',sol_16)

# Definir problema
Futoshiki7SemInferência = futoshiki(7, preenchidos7, desigualdades7)
# Guardar solucao
sol_17 = backtracking_search(Futoshiki7SemInferência)
# Calcular tempo medio executando 20 vezes
tempo_medio17 = avg_run_time(backtracking_search,Futoshiki7SemInferência)
# Guardar resultados
temposFutoshiki['Futoshiki7SemInferência'] = tempo_medio17
solucoesFutoshiki['Futoshiki7SemInferência'] = sol_17
# Imprimir resultados
print(f"Tempo médio de execução: {tempo_medio17}")
print('Assignment (Futoshiki7SemInferência) = ',sol_17)

# Definir problema
Futoshiki7ExtSemInferência = futoshiki(7, preenchidos7ext, desigualdades7ext)
# Guardar solucao
sol_17ext = backtracking_search(Futoshiki7ExtSemInferência)
# Calcular tempo medio executando 20 vezes
tempo_medio17ext = avg_run_time(backtracking_search,Futoshiki7ExtSemInferência)
# Guardar resultados
temposFutoshiki['Futoshiki7ExtSemInferência'] = tempo_medio17ext
solucoesFutoshiki['Futoshiki7ExtSemInferência'] = sol_17ext
# Imprimir resultados
print(f"Tempo médio de execução: {tempo_medio17ext}")
print('Assignment (Futoshiki7ExtSemInferência) = ',sol_17ext)

Tempo médio de execução: 0.0007503600005293265
Assignment (Futoshiki5SemInferência) =  {(0, 0): 3, (1, 0): 1, (2, 0): 2, (3, 0): 4, (4, 0): 5, (0, 1): 5, (1, 1): 4, (2, 1): 1, (3, 1): 2, (4, 1): 3, (0, 2): 1, (1, 2): 3, (2, 2): 4, (3, 2): 5, (4, 2): 2, (0, 3): 2, (1, 3): 5, (2, 3): 3, (3, 3): 1, (4, 3): 4, (0, 4): 4, (1, 4): 2, (2, 4): 5, (3, 4): 3, (4, 4): 1}
Tempo médio de execução: 0.0017226850031875074
Assignment (Futoshiki6SemInferência) =  {(0, 0): 1, (1, 0): 2, (2, 0): 5, (3, 0): 3, (4, 0): 4, (5, 0): 6, (0, 1): 6, (1, 1): 3, (2, 1): 2, (3, 1): 1, (4, 1): 5, (5, 1): 4, (0, 2): 5, (1, 2): 4, (2, 2): 1, (3, 2): 2, (4, 2): 6, (5, 2): 3, (0, 3): 2, (1, 3): 6, (2, 3): 3, (3, 3): 4, (4, 3): 1, (5, 3): 5, (0, 4): 4, (1, 4): 1, (2, 4): 6, (3, 4): 5, (4, 4): 3, (5, 4): 2, (0, 5): 3, (1, 5): 5, (2, 5): 4, (3, 5): 6, (4, 5): 2, (5, 5): 1}
Tempo médio de execução: 0.0015082550002261997
Assignment (Futoshiki7SemInferência) =  {(0, 0): 2, (1, 0): 7, (2, 0): 6, (3, 0): 3, (4, 0): 4, (5, 0): 1,

- Backtracking com inferência Forward Checking

In [136]:
# Definir problema
Futoshiki5ComInferência_FRWDchecking = futoshiki(5, preenchidos5, desigualdades5)
# Resolver e guardar solucao
sol_25 = backtracking_search(Futoshiki5ComInferência_FRWDchecking, inference=forward_checking)
# Calcular o tempo medio de execucao
tempo_medio25 = avg_run_time(backtracking_search, Futoshiki5ComInferência_FRWDchecking, inference=forward_checking)
# Guardar resultados
tempos['Futoshiki5ComInferência_FRWDchecking'] = tempo_medio25
solucoes['Futoshiki5ComInferência_FRWDchecking'] = sol_25
# Imprimir resultados
print(f"Tempo médio de execução: {tempo_medio25}")
print('Assignment (Futoshiki5ComInferência_FRWDchecking) = ',sol_25)

# Definir problema
Futoshiki6ComInferência_FRWDchecking = futoshiki(6, preenchidos6, desigualdades6)
# Resolver e guardar solucao
sol_26 = backtracking_search(Futoshiki6ComInferência_FRWDchecking, inference=forward_checking)
# Calcular o tempo medio de execucao
tempo_medio26 = avg_run_time(backtracking_search, Futoshiki6ComInferência_FRWDchecking, inference=forward_checking)
# Guardar resultados
tempos['Futoshiki6ComInferência_FRWDchecking'] = tempo_medio26
solucoes['Futoshiki6ComInferência_FRWDchecking'] = sol_26
# Imprimir resultados
print(f"Tempo médio de execução: {tempo_medio26}")
print('Assignment (Futoshiki6ComInferência_FRWDchecking) = ', sol_26)

# Definir problema
Futoshiki7ComInferência_FRWDchecking = futoshiki(7, preenchidos7, desigualdades7)
# Resolver e guardar solucao
sol_27 = backtracking_search(Futoshiki7ComInferência_FRWDchecking, inference=forward_checking)
# Calcular o tempo medio de execucao
tempo_medio27 = avg_run_time(backtracking_search, Futoshiki7ComInferência_FRWDchecking, inference=forward_checking)
# Guardar resultados
tempos['Futoshiki7ComInferência_FRWDchecking'] = tempo_medio27
solucoes['Futoshiki7ComInferência_FRWDchecking'] = sol_27
# Imprimir resultados
print(f"Tempo médio de execução: {tempo_medio27}")
print('Assignment (Futoshiki7ComInferência_FRWDchecking) = ', sol_27)

# Definir problema
Futoshiki7extComInferência_FRWDchecking = futoshiki(7, preenchidos7ext, desigualdades7ext)
# Resolver e guardar solucao
sol_27ext = backtracking_search(Futoshiki7extComInferência_FRWDchecking, inference=forward_checking)
# Calcular o tempo medio de execucao
tempo_medio27ext = avg_run_time(backtracking_search, Futoshiki7extComInferência_FRWDchecking, inference=forward_checking)
# Guardar resultados
tempos['Futoshiki7extComInferência_FRWDchecking'] = tempo_medio27ext
solucoes['Futoshiki7extComInferência_FRWDchecking'] = sol_27ext
# Imprimir resultados
print(f"Tempo médio de execução: {tempo_medio27ext}")
print('Assignment (Futoshiki7extComInferência_FRWDchecking) = ', sol_27ext)

Tempo médio de execução: 0.0008416299941018223
Assignment (Futoshiki5ComInferência_FRWDchecking) =  {(0, 0): 3, (1, 0): 1, (2, 0): 2, (3, 0): 4, (4, 0): 5, (0, 1): 5, (1, 1): 4, (2, 1): 1, (3, 1): 2, (4, 1): 3, (0, 2): 1, (1, 2): 3, (2, 2): 4, (3, 2): 5, (4, 2): 2, (0, 3): 2, (1, 3): 5, (2, 3): 3, (3, 3): 1, (4, 3): 4, (0, 4): 4, (1, 4): 2, (2, 4): 5, (3, 4): 3, (4, 4): 1}
Tempo médio de execução: 0.0013467600132571534
Assignment (Futoshiki6ComInferência_FRWDchecking) =  {(0, 0): 1, (1, 0): 2, (2, 0): 5, (3, 0): 3, (4, 0): 4, (5, 0): 6, (0, 1): 2, (1, 1): 6, (2, 1): 3, (3, 1): 1, (4, 1): 5, (5, 1): 4, (0, 2): 5, (1, 2): 3, (2, 2): 4, (3, 2): 2, (4, 2): 6, (5, 2): 1, (0, 3): 6, (1, 3): 5, (2, 3): 2, (3, 3): 4, (4, 3): 1, (5, 3): 3, (0, 4): 4, (1, 4): 1, (2, 4): 6, (3, 4): 5, (4, 4): 3, (5, 4): 2, (0, 5): 3, (1, 5): 4, (2, 5): 1, (3, 5): 6, (4, 5): 2, (5, 5): 5}
Tempo médio de execução: 0.001965565004502423
Assignment (Futoshiki7ComInferência_FRWDchecking) =  {(0, 0): 2, (1, 0): 3, (2, 0

## Visualização do problema

In [ ]:
# Implemente uma função que permita visualizar o puzzle Futoshiki, antes e depois de resolvido. Compare com a solução obtida pelo seu algoritmo. 
# No caso de não implementar esta função, inclua um screenshot do problema e da sua solução.
def visualizacao_futoshiki(n, problema_csp, desigualdades, solucao):
    
    def mapeamento_2D_para_1D(x,y):
        return y*(n+(n-1)*3+1) + x

    def escrita_numeros(x,y):
        return mapeamento_2D_para_1D(x*4, y*2)
    
    def escrita_desigualdades_linha(x,y):
        return mapeamento_2D_para_1D(x*4+2, y*2)
    
    def escrita_desigualdades_coluna(x,y):
        return mapeamento_2D_para_1D(x*4, y*2-1)

    def escrever_desigualdade(operacao, direcao):
        if direcao == "x":
            if operacao.__name__ == "maior":
                return '>'
            else:
                return '<'
        else:
            if operacao.__name__ == "maior":
                return u"\u2227" # Carater ∧
            else:
                return u"\u2228" # Carater ∨

    board_inicial = ([" "] * (n+(n-1)*3) + ["\n"]) * (n*2)
    for var, dom in problema_csp.domains.items():
        if len(dom) == 1:
            board_inicial[escrita_numeros(var[0],var[1])] = str(dom[0])
        else:
            board_inicial[escrita_numeros(var[0],var[1])] = "0"
    for var, operacao in desigualdades.items():
        if (var[0][1] == var[1][1]):
            board_inicial[escrita_desigualdades_linha(var[0][0],var[0][1])] = escrever_desigualdade(operacao, "x")
        else:
            board_inicial[escrita_desigualdades_coluna(var[0][0],var[0][1])] = escrever_desigualdade(operacao, "y")
    
    board_final = board_inicial.copy()
    for var, valor in solucao.items():
        board_final[escrita_numeros(var[0],var[1])] = str(valor)
    

    print("Estado Inicial:")
    print(''.join(board_inicial))
    print("Solução:")
    print(''.join(board_final))

# Implentação de ALgoritmos

- Backtracking sem inferência

In [ ]:
# Definir problema
QuadradoLatinoSemInferência = quadrado_latino(n)
# Guardar solucao
sol_1 = backtracking_search(QuadradoLatinoSemInferência)
# Calcular tempo medio executando 20 vezes
tempo_medio1 = avg_run_time(backtracking_search, QuadradoLatinoSemInferência)
# Guardar resultados
tempos['QuadradoLatinoSemInferência'] = tempo_medio1
solucoes['QuadradoLatinoSemInferência'] = sol_1
# Imprimir resultados
print(f"Tempo médio de execução: {tempo_medio1}")
print('Assignment (QuadradoLatinoSemInferência) = ',sol_1)